docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,157 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
 
--   See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
+- [Don’t panic after `await`](https://mmapped.blog/posts/01-effective-rust-canisters.html#panic-await) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
 
--   For context: [IC interface spec on message execution](../ic-interface-spec#message-execution)
+- See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
 
-### Be aware that state may change during inter-canister calls
+- For context: [IC interface spec on message execution](../ic-interface-spec#message-execution)
+
+### Be aware that there is no reliable message ordering
 
 #### Security Concern
 
-Messages (but not entire calls) are processed atomically. This can lead to security issues, such as:
+As described in the Message Execution Basics above, messages (but not entire calls) are processed atomically. This can lead to a variety of severe security issues such as:
+
+* _Time-of-check time-of-use issues._ These occur when some condition on global state is checked before an inter-canister call is made, and then wrongly assuming the condition to still hold when the call returns. For example, one might check if there is sufficient balance on some account, then issue two inter-canister calls (e.g. making transfers). When the second inter-canister call starts, it is possible that the condition checked initially no longer holds, because other ledger transfers may have happened before the callback of the first call is executed (see also Property 3 above). 
 
--   Time-of-check time-of-use: checking some condition on global state before an inter-canister call and wrongly assuming it to still hold when the call returned.
+* _Double-Spending issues._ Such issues occur when a transfer is issued twice, often because of unfavorable message scheduling. For example, suppose we check if a caller is eligible for a refund and if so, transfer some refund amount to them. When the refund ledger call returns successfully, we set a flag in the canister storage indicating that the caller has been refunded. This is vulnerable to double-spending because the refund method can be called twice by the caller in parallel, in which case it is possible that the messages before issuing the transfer (including the eligibility check) are scheduled before both callbacks. A detailed explanation of this issue can be found in the [community conversation on Security Best Practices](https://www.youtube.com/watch?v=PneRzDmf_Xw&list=PLuhDt1vhGcrez-f3I0_hvbwGZHZzkZ7Ng&index=2&t=4s).   
```

Comment:
Isnt't this the same as the time-of-check time-of-use example above?

Reply:
The roughly same class of bugs are usually dubbed "reentrancy bugs" in the Ethereum world. There are quite a few resources on this, so maybe makes sense to mention the connection here?

Reply:
@oggy-dfin 
> The roughly same class of bugs are usually dubbed "reentrancy bugs" in the Ethereum world. There are quite a few resources on this, so maybe makes sense to mention the connection here?

Makes sense. I extended the first paragraph to account for this. In case you have more links / references you think we should add, please LMK. 

@roelstorms 
> Isnt't this the same as the time-of-check time-of-use example above? It's an instance of such a vulnerability but it's (almost) the same as the example given above.

I agree they are quite similar. Still IMO given the severity of these issues (and since we are actually seeing both problems), I think it may be worth describing both. 

Also, there are some differences: The first issue aims at describing typical TOCTOU (check something, wrongly assume the condition still holds later). The second issue in my view is not exactly of this type, as the it occurs due to reentrancy to the same method directly because of updating the state too late. 

WDYT? If you see a nice way of improving this I'm all in :-)

Reply:
In the double spending issue you explain that the first step is to check eligibility for a refund.
Then you do the refund and thus the first check should now fail.

So basically, it's a TOC TOU issue but with two calls of the same code. While in general TOC TOU could be different calls affecting common state.

If you wouldn't "check" in the double spending example but retrieve some account balance, store it in a local variable, transfer it (all in separate messages), you would have a double spending issue without explicitly checking some state. So the two categories of problems would look less similar. But in the end, I agree with @oggy-dfin, these are instances of a broader category of vulnerabilities. Your change in the last commit clarifies this. Good for me!